brave/brave-ios
# Adding support for FIDO/FIDO2 APIs

fix https://github.com/brave/brave-ios/issues/1148
fix https://github.com/brave/brave-ios/issues/1150
fix https://github.com/brave/brave-ios/issues/1159
fix https://github.com/brave/brave-ios/issues/1163
fix https://github.com/brave/brave-ios/issues/1176

## Description
This PR implements javascript FIDO/FIDO2 APIs for the Brave iOS browser. With these APIs users will be able to use the lightning yubikeys keys as a second factor for authentication.

FIDO2 (WebAuthn) and FIDO rely on javascript APIs (navigator.credentials.create, navigator.credentials.get, u2f.sign, u2f.register). These are currently not available with WKWebView.

Scripts which define these APIs are injected at document start and end. These APIs hooks invoke services provided by the Yubico SDK to sign and register the requests passed using javascript.

For more details about the APIs refer to:
https://fidoalliance.org/fido2/
https://FIDOalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-javascript-api.html

## Submitter Checklist:


## Test Plan:

1. Navigate to the demo sites and make sure FIDO/FIDO2 workflow completes.
2. For other real sites:
  - create an account
  - enable 2FA
  - Registering 2FA keys should succeed.
  - Logout and try logging back in with the security key
  - Should succeed.

### Demo Sites:

1. WebAuthn
  - demo.yubico.com/u2f
  - webauthn.io
  - webauthntest.azurewebsites.net/
2. High Level Legacy U2F - mdp.github.io/u2fdemo
3. Low Level Legacy U2F - u2fdemo.hypersecu.com
4. APIs in secure context - http://dustiest-limitation.000webhostapp.com/nonsecure-u2f.html
5. APIs in 3p iframes - http://dustiest-limitation.000webhostapp.com/fido-iframe.html

### Sites that are working:
Without UA change:

1. github.com
2. bitbucket.org
3. 1Password.com
4. login.gov

Need UA change: https://github.com/brave/brave-ios/pull/1158/files#diff-1b680c574387c7d713ce005b70d406acR58

1. console.aws.amazon.com
2. twitter.com

## Reviewer Checklist:
